src/FlashLoanRouter.sol

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8;
+
+import {ICowSettlement} from "./interface/ICowSettlement.sol";
+import {IFlashLoanRouter} from "./interface/IFlashLoanRouter.sol";
+import {IFlashLoanSolverWrapper} from "./interface/IFlashLoanSolverWrapper.sol";
+import {BytesUtil, LoanRequest, LoansWithSettlement} from "./library/LoansWithSettlement.sol";
+import {ICowAuthentication} from "./vendored/ICowAuthentication.sol";
+import {IERC20} from "./vendored/IERC20.sol";
+
+contract FlashLoanRouter is IFlashLoanRouter {
+    using BytesUtil for bytes;
+    using LoansWithSettlement for bytes;
+
+    IFlashLoanSolverWrapper NO_PENDING_BORROWER = IFlashLoanSolverWrapper(address(0));
+
+    ICowSettlement public immutable settlementContract;
+    ICowAuthentication public immutable settlementAuthentication;
+
+    IFlashLoanSolverWrapper private transient pendingBorrower;
+    bytes32 private transient pendingDataHash;
+
+    modifier onlySolver() {
+        // Todo: investigate security implication of self calls.
+        require(settlementAuthentication.isSolver(msg.sender), "Only callable by a solver");
+        _;
+    }
+
+    modifier onlyPendingBorrower() {
+        // Todo: investigate security implication of self calls.
+        require(msg.sender == address(pendingBorrower), "Only callable by borrower");
+        _;
+    }
+
+    constructor(ICowSettlement _settlementContract) {
+        settlementContract = _settlementContract;
+        settlementAuthentication = ICowAuthentication(_settlementContract.authenticator());
+    }
+
+    function flashLoanAndSettle(LoanRequest.Data[] calldata loans, bytes calldata settlement) external onlySolver {
+        bytes memory encodedLoansWithSettlement = LoansWithSettlement.encodeLoansWithSettlement(loans, settlement);
```

Comment:
This library is indeed quite involved. If we used `abi.encode(loans, settlement, index)` instead of popping in place and manually decoding the loans and settlement would we have a lot more gas overhead?

Also, are we sure the tstore approach is not more gas efficient here (it seems conceptually a bit simpler)